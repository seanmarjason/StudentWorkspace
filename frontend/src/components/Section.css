.section {
  min-height: 15vh;
  width: 80vw;
  margin: 5vh 10vw;
  padding: 30px;
  box-shadow: 0 0 25px rgb(0 0 0 / 20%);
}

.sectionArtefacts {
  display: flex;
  flex-wrap: wrap;
  height: 100%;
  width: 95%;
  position: relative;
}

.section h3 {
  width: 100%;
  margin: 0 0 10px 0;
  text-transform: capitalize;
}

.artefact {
  background-color: lightgray;
  border: 1px solid lightgray;
  border-radius: 3px;
  text-align: center;
  cursor: pointer;
  display: inline-block;
  height: 10vh;
  width: 8vh;
  margin: 10px;
  overflow: hidden;
}

.artefact:hover {
  border: 1px solid rgb(86, 86, 255);
}

.addArtefact {
  position: absolute;
  right: 0;
  display: inline-block;
  cursor: pointer;
  border-radius: 3px;
  height: 10vh;
  width: 8vh;
  margin: 10px;
}

.artefactLink {
  text-decoration: none;
  color: black;
  width: 100%;
  height: 100%;
}
