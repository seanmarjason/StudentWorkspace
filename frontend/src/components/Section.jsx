import { useState, useEffect } from 'react';
import axios from 'axios';
import './Section.css'
import { UploadForm } from './UploadForm';
import { Artefact } from './Artefact';

const Section = ({ groupId, sectionName }) => {

  const [showUploadForm, setShowUploadForm] = useState(false);
  const [sectionData, setSectionData] = useState([]);
  const [artefact, setArtefact] = useState(null);

  useEffect(() => {
    const url = `http://localhost:3000/documents/list/${groupId.toString()}/${sectionName}`

    axios.get(url)
      .then(response => setSectionData(response.data.documents))
      .catch(error => console.log(`Error: ${error}`))
  }, [groupId, sectionName, showUploadForm]);

  const handleArtefactClick = (event) => {
    const artefact = JSON.parse(event.target.value);
    setArtefact(artefact)
  }

  return (
    <div className="section">
      <h3>{sectionName}</h3>
      <div className="sectionArtefacts">
        {
          sectionData.map(artefact => 
              <button
                className="artefact"
                value={JSON.stringify(artefact)}
                onClick={(event) => handleArtefactClick(event)}
              >
                {artefact}
              </button>
          )
        }
        <button 
          className="addArtefact"
          onClick={() => setShowUploadForm(true)}
        >
          Add Artefact
        </button>
        {
        showUploadForm && 
          <UploadForm
            section={sectionName}
            callback={setShowUploadForm}
          />
        }
        {
        artefact &&
          <Artefact                 
            artefact={artefact}
            setArtefact={setArtefact}
          />
        }
      </div>
    </div>
  )
};

export { Section }
