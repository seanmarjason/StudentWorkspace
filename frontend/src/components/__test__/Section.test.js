import {render, screen, waitFor, fireEvent } from '@testing-library/react';
import axios from 'axios';
import { Section } from '../Section';

const sectionWithNoDocuments = {
  data: {
    documents: []
  }
}

const sectionWithDocuments = {
  data: {
    documents: [
      'doc 1',
      'doc 2',
      'doc 3',
    ]
  }
}

jest.mock('axios');

jest.mock('../UploadForm', () => ({
  UploadForm: () => <div className="uploadForm">UploadForm</div>
  }
));

jest.mock('../Artefact', () => ({
  Artefact: () => <div className="artefact"><h3>Artifact</h3></div>
  }
));

describe('Section' , () => {
  afterEach(() =>
    jest.clearAllMocks()
  )

  it('should match snapshot with no documents', async () => {
    axios.get.mockResolvedValueOnce(sectionWithNoDocuments);

    const view = render(<Section groupId={'100'} sectionName="deliverables" />)

    await waitFor(() => {
      screen.getByText('deliverables')
    });

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot with documents', async () => {
    axios.get.mockResolvedValueOnce(sectionWithDocuments);

    const view = render(<Section groupId={'100'} sectionName="deliverables" />)

    await waitFor(() => {
      screen.getByText('doc 1')
    });

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when user clicks an artifact', async () => {
    axios.get.mockResolvedValue(sectionWithDocuments);

    const view = render(<Section groupId={'100'} sectionName="deliverables" />)

    await waitFor(() => {
      screen.getByText('doc 1')
    });

    fireEvent(
      screen.getByText('doc 1'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    )

    await waitFor(() => {
      screen.getByText('Artifact')
    });
    
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when user tries to add artifact', async () => {
    axios.get.mockResolvedValue(sectionWithDocuments);

    const view = render(<Section groupId={'100'} sectionName="deliverables" />)

    await waitFor(() => {
      screen.getByText('Add Artefact')
    });

    fireEvent(
      screen.getByText('Add Artefact'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    )

    await waitFor(() => {
      screen.getByText('UploadForm')
    });
    
    expect(view.asFragment()).toMatchSnapshot();
  });
  
})
  