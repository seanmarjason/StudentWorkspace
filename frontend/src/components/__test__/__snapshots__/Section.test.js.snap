// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section should match snapshot when user clicks an artifact 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <h3>
      deliverables
    </h3>
    <div
      class="sectionArtefacts"
    >
      <button
        class="artefact"
        value="\\"doc 1\\""
      >
        doc 1
      </button>
      <button
        class="artefact"
        value="\\"doc 2\\""
      >
        doc 2
      </button>
      <button
        class="artefact"
        value="\\"doc 3\\""
      >
        doc 3
      </button>
      <button
        class="addArtefact"
      >
        Add Artefact
      </button>
      <div
        class="artefact"
      >
        <h3>
          Artifact
        </h3>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Section should match snapshot when user tries to add artifact 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <h3>
      deliverables
    </h3>
    <div
      class="sectionArtefacts"
    >
      <button
        class="artefact"
        value="\\"doc 1\\""
      >
        doc 1
      </button>
      <button
        class="artefact"
        value="\\"doc 2\\""
      >
        doc 2
      </button>
      <button
        class="artefact"
        value="\\"doc 3\\""
      >
        doc 3
      </button>
      <button
        class="addArtefact"
      >
        Add Artefact
      </button>
      <div
        class="uploadForm"
      >
        UploadForm
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Section should match snapshot with documents 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <h3>
      deliverables
    </h3>
    <div
      class="sectionArtefacts"
    >
      <button
        class="artefact"
        value="\\"doc 1\\""
      >
        doc 1
      </button>
      <button
        class="artefact"
        value="\\"doc 2\\""
      >
        doc 2
      </button>
      <button
        class="artefact"
        value="\\"doc 3\\""
      >
        doc 3
      </button>
      <button
        class="addArtefact"
      >
        Add Artefact
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Section should match snapshot with no documents 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <h3>
      deliverables
    </h3>
    <div
      class="sectionArtefacts"
    >
      <button
        class="addArtefact"
      >
        Add Artefact
      </button>
    </div>
  </div>
</DocumentFragment>
`;
